core/src/context.rs

```diff
@@ -2,21 +2,120 @@ use error::HolochainError;
 use holochain_agent::Agent;
 use logger::Logger;
 use persister::Persister;
-use std::sync::{Arc, Mutex};
+use state::State;
+use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard};
 
-/// Context holds those aspects of the outside world that a Holochain instance needs to operate
+/// Context holds those aspects of a living Holochain instance that components need to operate.
+/// This includes modules that are injected from the outside like logger and persister
+/// but also the state of the instance that gets injected before passing on the context
+/// to inner components/reducers.
 #[derive(Clone)]
 pub struct Context {
     pub agent: Agent,
     pub logger: Arc<Mutex<Logger>>,
     pub persister: Arc<Mutex<Persister>>,
+    state: Option<Arc<RwLock<State>>>,
 }
 
 impl Context {
+    pub fn new(
+        agent: Agent,
+        logger: Arc<Mutex<Logger>>,
+        persister: Arc<Mutex<Persister>>,
+    ) -> Context {
+        Context {
+            agent,
+            logger,
+            persister,
+            state: None,
+        }
+    }
     // helper function to make it easier to call the logger
     pub fn log(&self, msg: &str) -> Result<(), HolochainError> {
         let mut logger = self.logger.lock().or(Err(HolochainError::LoggingError))?;
         logger.log(msg.to_string());
         Ok(())
     }
+
+    pub fn set_state(&mut self, state: Arc<RwLock<State>>) {
```

Comment:
Use something like pub(self) so its not public to everyone, as intended from PR description.

Reply:
Since I use it in Instance, I can only restrict it to the crate level: `pub(crate)`.
But yes.